pulsar-broker/src/main/java/org/apache/pulsar/broker/intercept/ManagedLedgerInterceptorImpl.java

```diff
@@ -131,4 +153,47 @@ public void onUpdateManagedLedgerInfo(Map<String, String> propertiesMap) {
             }
         }
     }
+
+    private PayloadProcessorHandle processPayload(Set<ManagedLedgerPayloadProcessor.Processor> processors,
+                                                  Object context, ByteBuf payload) {
+
+        ByteBuf tmpData = payload;
+        final Set<ImmutablePair<ManagedLedgerPayloadProcessor.Processor, ByteBuf>> processedSet = new LinkedHashSet<>();
+        for (ManagedLedgerPayloadProcessor.Processor payloadProcessor : processors) {
+            if (payloadProcessor != null) {
+                tmpData = payloadProcessor.process(context, tmpData);
+                processedSet.add(new ImmutablePair<>(payloadProcessor, tmpData));
+            }
+        }
+        final ByteBuf dataToReturn = tmpData;
+        return new PayloadProcessorHandle() {
+            @Override
+            public ByteBuf getProcessedPayload() {
+                return dataToReturn;
+            }
+
+            @Override
+            public void release() {
+                for (ImmutablePair<ManagedLedgerPayloadProcessor.Processor, ByteBuf> p : processedSet) {
+                    p.left.release(p.right);
```

Comment:
different payloadProcessor referenced to the same ByteBuf. And the same ByteBuf will be released multi times.
Additionally, we only need to record the final ByteBuf.

Reply:
I have fixed the comments in ManagedLedgerPayloadProcessor::Processor::release to explicitly state that the release is to be done only when process() returned a different ByteBuf instance.
LinkedHashSet is used for the processor set and hence the traversal order is the same as insertion order. The final ByteBuf is indeed recorded